rfcs/2021-06-15-shipping-blade.md

```diff
@@ -0,0 +1,355 @@
+# 🛳 Shipping Blade <!-- omit in toc -->
+
+### Table Of Contents <!-- omit in toc -->
+- [Summary](#summary)
+- [Basic Example](#basic-example)
+- [Motivation](#motivation)
+- [Detailed Design](#detailed-design)
+  - [What do we want to actually ship?](#what-do-we-want-to-actually-ship)
+  - [How will the consumers import things from blade?](#how-will-the-consumers-import-things-from-blade)
+    - [1. `import` individual components](#1-import-individual-components)
+      - [Pros](#pros)
+      - [Cons](#cons)
+    - [2. `import` everything from root package](#2-import-everything-from-root-package)
+      - [Pros](#pros-1)
+      - [Cons](#cons-1)
+    - [3. `import` from each category(components, tokens,types, utils) root package](#3-import-from-each-categorycomponents-tokenstypes-utils-root-package)
+      - [Pros](#pros-2)
+      - [Cons](#cons-2)
+    - [4. `import` named components from each category(components, tokens,types, utils) root package](#4-import-named-components-from-each-categorycomponents-tokenstypes-utils-root-package)
+      - [Pros](#pros-3)
+      - [Cons](#cons-3)
+    - [Which approach we'll take?](#which-approach-well-take)
+  - [What strategy to take for packaging?](#what-strategy-to-take-for-packaging)
+    - [1. Multi-package](#1-multi-package)
+    - [2. Mono-package](#2-mono-package)
+    - [Conclusion](#conclusion)
+  - [How to package?](#how-to-package)
+    - [Shall we Transpile(using babel)?](#shall-we-transpileusing-babel)
+    - [Shall we Bundle(using rollup/similar tools)?](#shall-we-bundleusing-rollupsimilar-tools)
+    - [Conclusion](#conclusion-1)
+  - [What module system to target?](#what-module-system-to-target)
+    - [CommonJS(a.k.a cjs)](#commonjsaka-cjs)
+    - [ES Modules(a.k.a esm)](#es-modulesaka-esm)
+    - [Conclusion](#conclusion-2)
+  - [Meta information in package.json](#meta-information-in-packagejson)
+- [Drawbacks/Constraints](#drawbacksconstraints)
+- [Alternatives](#alternatives)
+- [Adoption strategy](#adoption-strategy)
+- [Open Questions](#open-questions)
+- [References](#references)
+
+# Summary
+
+We can't ship the raw code to the consumers so we need to ship/publish blade. But to make it publishable there are certain things we need to keep in mind and answer relevant questions.
+
+# Basic Example
+We should be able to install and use blade in our code without any overhead.
+```js
+import { Button, Text } from '@razorpay/blade/components'
+import { overrideTheme, paymentTheme } from '@razorpay/blade/theme'
+import { getColorScheme, useColorScheme } from '@razorpay/blade/utils'
+```
+
+# Motivation
+- We can't ship raw source code to our consumers since we are not aware about the build tool chains they use, what config do they support.
+- We need to prevent access to certain parts of the codebase. There might be certain files/functionality which are just used for internal purpose and we don't want to expose them to our consumers. So we want to publish few things and not publish few.
```

Comment:
Can we solve this con by having [no-restricted-imports ESLint rule](https://eslint.org/docs/rules/no-restricted-imports)? The consumers may have to add this to their rules though so might be some extra work for them.

Reply:
umm eslint might only prevent but if people want to access they'll do it regardless by turning off the rule or disabling at that line 🙈